### ISSUE TITLE
No bitwise flag operations ?

### ISSUE THREAD

**Issue by emanueljelp** [realm/realm-java#1949]
How can I filter a hex integer flag to realm results?

**Comment by cmelchior** [comment]
Can you give an example of what you want to achieve?

**Comment by emanueljelp** [comment]
from an int value in the database by example in c
int flags =0x0;

#define DIRTY	0x01
#define OPEN	0x02
#define VERBOSE	0x04
#define RED	0x08
#define SEASICK	0x10

flags = flags | DIRTY;		/* set DIRTY bit */
flags = flags & ~DIRTY;		/* clear DIRTY bit */

then I could find in my table if(flags & DIRTY) , if(!(flags & OPEN)), etc. with only one variable.

**Comment by cmelchior** [comment]
I am still not sure what you want to achieve. You can store integers in Realm, and nothing prevents you doing bitwise operations on the integer you want to search for?

```
// If Foo.field is an int
realm.where(Foo.class).equalTo("field", flags & DIRTY).findAll()
```

**Comment by emanueljelp** [comment]
realm.where(Foo.class).equalTo("field",field_value & DIRTY).findAll(); from where you get flags value??

**Comment by emanueljelp** [comment]
I will need other query to get all values??

**Comment by emanueljelp** [comment]
here the detailed example I want to do :/, maybe I'm wrong sorry but I don't want to do a string search.
http://www.eskimo.com/~scs/cclass/int/sx4ab.html
I'm seeing the realm code, maybe i can do this implementation and do a pull request. My problem is the time :(
Regards.

**Comment by kneth** [comment]
Realm's storage/query engine doesn't support bitwise operators yet. Currently, you will have to iterate through your objects to do it:

    for(Foo foo : realm.allObjects(Foo.class) {
        if (foo.getField() & DIRTY) {
           // DIRTY is set and do something
        }
    }

Realm is storing booleans efficiently (see https://realm.io/news/jp-simard-realm-core-database-engine/ - roughly 14:18 minutes in the video) so it might be better to have easy flag as a separate field. By doing so, you have the full speed and elegancy of Realm's query engine.

**Comment by emanueljelp** [comment]
Thanks :), the boolean's way sounds good, but I will need migrate my database every time I add a new value :/ but by now will works. The iteration way, doesn't like me, the point here is optimize my query, and  I can do a string query in that case. Thanks again.

**Comment by kneth** [comment]
I'm not quite sure what you mean by string query. To be more concrete, my proposal was to have a model class like:

    class Foo extends RealmObjects {
        private boolean dirty;
        private boolean open;
        // setters and getters left out
    }

and now your query can be:

    RealmResults<Foo> foos = realm.where(Foo.class).equalTo("dirty", true).findAll();

**Comment by emanueljelp** [comment]
Oh, sorry my english is very bad. I said that if my backend set's a new section to filter, I can find in the database a tag in a string. Or a value in a json string saved in the database. But your boolean way sounds ok, I can set several booleans not defined or an array of booleans that is basically the same thing that the entire value.

**Comment by kneth** [comment]
If your data is coming from a JSON string, you'll have to transform it into a set of boolean values. Storing a boolean (just one bit, right?) is very space efficient. And queries are pretty simple to write e.g., if you wish to query when two flags are set:

    RealmResults<Foo> foos = realm.where(Foo.class).equalTo("dirty", true).equalTo("open", true).findAll();

**Comment by CriseX** [comment]
@kneth are there any plans to add support for bitwise operations in the future?

Bitmasks are nifty because you can send them over the wire without need for serialization on either end unlike the boolean based approach you suggest here (which would require conversion to json or something, that is unnecessarily verbose)

**Comment by MkazemAkhgary** [comment]
nah, instead of just bit-wise operations realm should give us freedom by filtering collections using a predicated. i don't see why that's not supported. maybe because you can do pretty much anything in a predicate and cause undefined behavior or something. @CriseX
